pkg/virt-handler/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,131 @@
+package filewatcher
+
+import (
+	"errors"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type Event uint32
+
+const (
+	Create Event = 1 << iota
+	Remove
+	InoChange
+)
+
+type FileWatcher struct {
+	path     string
+	interval time.Duration
+
+	Events chan Event
+	Errors chan error
+	done   chan struct{}
+
+	lastIno uint64
+	closeMu sync.Mutex
+}
+
+func New(path string, interval time.Duration) *FileWatcher {
+	return &FileWatcher{
+		path:     path,
+		interval: interval,
+		Events:   make(chan Event),
+		Errors:   make(chan error),
+		done:     make(chan struct{}),
+	}
+}
+
+func (f *FileWatcher) Run() {
+	f.statFirst()
+	go func() {
+		defer close(f.Events)
+		defer close(f.Errors)
+
+		ticker := time.Tick(f.interval)
+		for {
+			select {
+			case <-f.done:
+				return
+			case <-ticker:
+				f.stat()
+			}
+		}
+	}()
+}
+
+func (f *FileWatcher) Close() {
+	f.closeMu.Lock()
+	if f.IsClosed() {
+		f.closeMu.Unlock()
+		return
+	}
+	close(f.done)
+	f.closeMu.Unlock()
+	return
+}
```

Comment:
Rather than having `Close()` it be more canonical go to simply pass a done chan to `New()` and have the caller close that?

Reply:
`Close()` is required here because on teardown of virt-handler we need to wait for the monitor to finish with cleaning up. That is similar to how it is down with heartbeat, which also returns a channel `heartBeatDone` from its `Run` method. However, I didn't find this design particularly easy to understand.

If you want I can change it though.

Reply:
@mhenriks if we do it in a follow up? Same for the other comment?

Reply:
That's fine especially if there are a lot of changes required that could be problmatic with backporting. I assume this pr will be backported